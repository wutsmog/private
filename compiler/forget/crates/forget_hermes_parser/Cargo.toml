[package]
name = "forget_hermes_parser"
version = "0.1.0"
authors.workspace = true
description.workspace = true
edition.workspace = true
homepage.workspace = true
keywords.workspace = true
license.workspace = true
repository.workspace = true

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
insta = { workspace = true }
forget_diagnostics = { workspace = true }
forget_estree = { workspace = true }
hermes = { workspace = true } 
juno_support = { workspace = true }
serde_json = { workspace = true }

[build-dependencies]
forget_estree_codegen = { workspace = true }