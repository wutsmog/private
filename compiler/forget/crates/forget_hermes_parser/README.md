# forget_hermes_parser

Wrapper around the Hermes parser that exposes parse results as a `forget_estree` AST.