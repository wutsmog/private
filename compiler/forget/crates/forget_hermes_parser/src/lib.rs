mod generated;
mod generated_extension;

use forget_diagnostics::Diagnostic;
use forget_estree::Program;
use generated_extension::{Context, FromHermes};
use hermes::parser::{HermesParser, ParserDialect, ParserFlags};
use hermes::utf::utf8_with_surrogates_to_string;
use juno_support::NullTerminatedBuf;

pub fn parse(source: &str, _file: &str) -> Result<Program, Vec<Diagnostic>> {
    let buf = NullTerminatedBuf::from_str_check(source);
    let result = HermesParser::parse(
        ParserFlags {
            dialect: ParserDialect::TypeScript,
            enable_jsx: true,
            store_doc_block: true,
            strict_mode: true,
        },
        &buf,
    );
    let mut cx = Context::new();
    if result.has_errors() {
        let error_messages = result.messages();
        return Err(error_messages
            .into_iter()
            .map(|diag| {
                let message = utf8_with_surrogates_to_string(diag.message.as_slice()).unwrap();
                Diagnostic::invalid_syntax(message, None)
            })
            .collect());
    }

    Ok(FromHermes::convert(&mut cx, result.root().unwrap()))
}
