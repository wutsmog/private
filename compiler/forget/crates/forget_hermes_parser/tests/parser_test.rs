use std::env;

use forget_hermes_parser::parse;
use insta::{assert_snapshot, glob};
use serde_json;

#[test]
fn fixtures() {
    glob!("fixtures/**.js", |path| {
        println!("fixture {}", path.to_str().unwrap());
        let input = std::fs::read_to_string(path).unwrap();
        let ast = parse(&input, path.to_str().unwrap()).unwrap();
        let output = serde_json::to_string_pretty(&ast).unwrap();
        let output = output.trim();
        assert_snapshot!(format!("Input:\n{input}\n\nOutput:\n{output}"));
    });
}
