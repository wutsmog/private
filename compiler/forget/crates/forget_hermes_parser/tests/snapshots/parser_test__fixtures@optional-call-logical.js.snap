---
source: crates/forget_hermes_parser/tests/parser_test.rs
expression: "format!(\"Input:\\n{input}\\n\\nOutput:\\n{output}\")"
input_file: crates/forget_hermes_parser/tests/fixtures/optional-call-logical.js
---
Input:
function Component(props) {
  const item = useFragment(graphql`...`, props.item);
  return item.items?.map((item) => renderItem(item)) ?? [];
}


Output:
{
  "type": "Program",
  "body": [
    {
      "type": "FunctionDeclaration",
      "id": {
        "type": "Identifier",
        "name": "Component",
        "typeAnnotation": null,
        "loc": null,
        "range": {
          "start": 9,
          "end": 18
        }
      },
      "params": [
        {
          "type": "Identifier",
          "name": "props",
          "typeAnnotation": null,
          "loc": null,
          "range": {
            "start": 19,
            "end": 24
          }
        }
      ],
      "body": {
        "type": "BlockStatement",
        "body": [
          {
            "type": "VariableDeclaration",
            "kind": "const",
            "declarations": [
              {
                "type": "VariableDeclarator",
                "id": {
                  "type": "Identifier",
                  "name": "item",
                  "typeAnnotation": null,
                  "loc": null,
                  "range": {
                    "start": 36,
                    "end": 40
                  }
                },
                "init": {
                  "type": "CallExpression",
                  "callee": {
                    "type": "Identifier",
                    "name": "useFragment",
                    "typeAnnotation": null,
                    "loc": null,
                    "range": {
                      "start": 43,
                      "end": 54
                    }
                  },
                  "arguments": [
                    {
                      "type": "TaggedTemplateExpression",
                      "tag": {
                        "type": "Identifier",
                        "name": "graphql",
                        "typeAnnotation": null,
                        "loc": null,
                        "range": {
                          "start": 55,
                          "end": 62
                        }
                      },
                      "quasi": {
                        "type": "TemplateLiteral",
                        "quasis": [
                          {
                            "type": "TemplateElement",
                            "tail": true,
                            "value": {
                              "cooked": "...",
                              "raw": "..."
                            },
                            "loc": null,
                            "range": {
                              "start": 62,
                              "end": 67
                            }
                          }
                        ],
                        "expressions": [],
                        "loc": null,
                        "range": {
                          "start": 62,
                          "end": 67
                        }
                      },
                      "loc": null,
                      "range": {
                        "start": 55,
                        "end": 67
                      }
                    },
                    {
                      "type": "MemberExpression",
                      "object": {
                        "type": "Identifier",
                        "name": "props",
                        "typeAnnotation": null,
                        "loc": null,
                        "range": {
                          "start": 69,
                          "end": 74
                        }
                      },
                      "property": {
                        "type": "Identifier",
                        "name": "item",
                        "typeAnnotation": null,
                        "loc": null,
                        "range": {
                          "start": 75,
                          "end": 79
                        }
                      },
                      "computed": false,
                      "loc": null,
                      "range": {
                        "start": 69,
                        "end": 79
                      }
                    }
                  ],
                  "loc": null,
                  "range": {
                    "start": 43,
                    "end": 80
                  }
                },
                "loc": null,
                "range": {
                  "start": 36,
                  "end": 80
                }
              }
            ],
            "loc": null,
            "range": {
              "start": 30,
              "end": 81
            }
          },
          {
            "type": "ReturnStatement",
            "argument": {
              "type": "LogicalExpression",
              "operator": "??",
              "left": {
                "type": "OptionalCallExpression",
                "callee": {
                  "type": "OptionalMemberExpression",
                  "object": {
                    "type": "MemberExpression",
                    "object": {
                      "type": "Identifier",
                      "name": "item",
                      "typeAnnotation": null,
                      "loc": null,
                      "range": {
                        "start": 91,
                        "end": 95
                      }
                    },
                    "property": {
                      "type": "Identifier",
                      "name": "items",
                      "typeAnnotation": null,
                      "loc": null,
                      "range": {
                        "start": 96,
                        "end": 101
                      }
                    },
                    "computed": false,
                    "loc": null,
                    "range": {
                      "start": 91,
                      "end": 101
                    }
                  },
                  "property": {
                    "type": "Identifier",
                    "name": "map",
                    "typeAnnotation": null,
                    "loc": null,
                    "range": {
                      "start": 103,
                      "end": 106
                    }
                  },
                  "computed": false,
                  "optional": true,
                  "loc": null,
                  "range": {
                    "start": 91,
                    "end": 106
                  }
                },
                "arguments": [
                  {
                    "type": "ArrowFunctionExpression",
                    "id": null,
                    "params": [
                      {
                        "type": "Identifier",
                        "name": "item",
                        "typeAnnotation": null,
                        "loc": null,
                        "range": {
                          "start": 108,
                          "end": 112
                        }
                      }
                    ],
                    "body": {
                      "type": "CallExpression",
                      "callee": {
                        "type": "Identifier",
                        "name": "renderItem",
                        "typeAnnotation": null,
                        "loc": null,
                        "range": {
                          "start": 117,
                          "end": 127
                        }
                      },
                      "arguments": [
                        {
                          "type": "Identifier",
                          "name": "item",
                          "typeAnnotation": null,
                          "loc": null,
                          "range": {
                            "start": 128,
                            "end": 132
                          }
                        }
                      ],
                      "loc": null,
                      "range": {
                        "start": 117,
                        "end": 133
                      }
                    },
                    "generator": true,
                    "async": false,
                    "loc": null,
                    "range": {
                      "start": 107,
                      "end": 133
                    },
                    "expression": true,
                    "loc": null,
                    "range": {
                      "start": 107,
                      "end": 133
                    }
                  }
                ],
                "optional": false,
                "loc": null,
                "range": {
                  "start": 91,
                  "end": 134
                }
              },
              "right": {
                "type": "ArrayExpression",
                "elements": [],
                "loc": null,
                "range": {
                  "start": 138,
                  "end": 140
                }
              },
              "loc": null,
              "range": {
                "start": 91,
                "end": 140
              }
            },
            "loc": null,
            "range": {
              "start": 84,
              "end": 141
            }
          }
        ],
        "loc": null,
        "range": {
          "start": 26,
          "end": 143
        }
      },
      "generator": false,
      "async": false,
      "loc": null,
      "range": {
        "start": 0,
        "end": 143
      },
      "loc": null,
      "range": {
        "start": 0,
        "end": 143
      }
    }
  ],
  "sourceType": "script",
  "loc": null,
  "range": {
    "start": 0,
    "end": 143
  }
}
